manual loading and parsing of dotenv in watch mode #558

tandibar opened the pull request:
I think this could be the fix for the issue #558 that I opened.

The problem with `dotenv.config().parsed` is that `dotenv.config()` modifies the `process.env`. So the fix uses `dotenv.parse()` directly so that `process.env is untouched.

All tests are green. I will provide an additional test in the next days.
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

#### Checklist

- [x] run `npm run test` ~and `npm run benchmark`~
- [x] tests and/or benchmarks are included
- [ ] ~documentation is changed or added~
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

tandibar committed:
manual loading and parsing of dotenv in watch mode #558

mcollina reviewed (approved):
lgtm

mcollina commented:
Thanks for opening a PR! Can you please add a unit test?

mcollina reviewed (changes requested):
Thanks for opening a PR! Can you please add a unit test?

tandibar committed:
added unit-test for #558

tandibar commented:
So I added a unit test and implemented the feature. The problem is, that it breaks another feature, see test `test/start.test.js should reload the env on restart when watching`. In the current implementation I think it is very hard to have both features. For me it would be more consistent if dotenv does not change its behavior in watch mode. But that would mean, that you have to restart your server if you change your `.env` even in watch mode.
What do you think?

mcollina commented:
> that would mean, that you have to restart your server if you change your .env even in watch mode.

Why? I _think_ it might be possible to have the env variables defined in process.env to have priority over the ones in .env.
Or viceversa. I think the surprising behavior is lack of consistency.

tandibar committed:
analyse which env vars should not be overwritten #558

tandibar commented:
So the problem ist, that `dotenv` is already loaded before, so we cannot just check which env vars are already present, because all are present. So I decided to check if there is a difference between the `process.env` and the parsed dotenv. If yes, we should not overwrite the env var. If it is the same, it seems to be managed by dotenv file, so we should overwrite.
Tests are now all passing, so I think that is the solution.
